WIP: Batch processing of individual msg types

Georacer opened the pull request:
Okay, it took the the whole evening, but I think I got something.

Comments inline, start from LogMsgGroup.

This is a work in progress, do not merge.

66.BIN finishes in 10 secs

Georacer committed:
WIP: Batch processing of individual msg types

66.BIN finishes in 10 secs

Georacer commented:
ReadLogLine(), findMsgStart(), readLogData() and trimTail() are no longer needed. Can we delete them?

Georacer commented:
One more thing: I'm worried that typecast may mess up data conversion due to little/big-endian difference between data.

In case we see odd figures, we should keep that in mind.

Georacer committed:
Catched potential out-of-bounds access of log

Georacer commented:
Just parsed the whole 160MB 51.BIN. Took 63.1 secs ^^

Georacer commented:
There is definitely something wrong with encoding using typecast now. I'll investigate more.

Georacer commented:
There is definitely something wrong with the exported data. They just don't make sense. I'll try to investigate more.

Georacer committed:
Fixed bug in usage of reshape

Georacer commented:
Okay, I think the last commit fixed it.
What was happening was that I was building an array for each field
A = [msg1_byte1 msg1_byte2...; msg2_byte1 msg2_byte2...] etc
and wanted to concatenate all rows in a single row, to give it to typecast (whcih accepts vectors only)
However, I used A(:) to do that, which gives
ans = [msg1_byte1 msg2_byte1...

hunt0r commented on Ardupilog.m:
```diff
@@ -83,52 +84,136 @@ function delete(obj)
+            % Read the FMT message
+            % Generate the index mask which corresponds to the places where
+            % FMT information exists
+            indices = reshape(ones(length(FMTIndices),1)*(3:(FMTLength-1)) + (ones(length(FMTIndices),1).*FMTIndices')*ones(1,FMTLength-3),[1 length(FMTIndices)*(FMTLength-3)]);
```
This line is hard for me to parse and understand. If it's okay with you, I might break it into a few lines or add more comments in a future commit.

Georacer committed:
Made generation of message data array more explicit

Georacer commented:
Okay, the latest commit fleshed the procedure out a bit. Thankfully there doesn't seem to be a hit in performance.

Georacer committed:
Stored all header index candidates

Now the log file is searched only once for candidate header positions.
Afterwards these are stored and each time we look for a message we start from there

Georacer commented:
Stored header indices so that we don't need to search the whole log for each message.
This led to 80% decrease in time.
Now the 51.BIN file is parsed in 15 secs!

hunt0r commented:
Excellent work!

Georacer commented:
Can I delete the now unused functions?

hunt0r commented:
Yes. I think you can also pull out the large commented block, and any other relics from the previous approach.

Some minor ideas to consider:
1) Is it worth it to functionalize the copied portions of code for organizing and then storing the message data? If I'm not mistaken we call it in one spot for FMT messages, and identically again for the rest of the messages.
2) Do you know the reshape() function? This is more natural to me than ones()*[values] for forming a matrix of repeated values, but does the same thing. (Not worth changing if you are used to the ones()*[values] syntax, I figured it out with a pen and paper)

Georacer commented:
I used reshape a few times in this PR. But I can't think of a way to use it to replace ones(r,c), when I need to create some arrays here and there.

Anyway, I'll create another PR for the cleanup tomorrow or Monday.
Then you can suggest changes if you want.

I'll also add another PR afterwards for choosing the desired message set at initializaiton.

hunt0r commented:
Oh damn! Sorry, I meant "repmat()"  I also didn't mean for my multiplication signs to start/stop italics... I gotta get better at github. (I must have had reshape on my mind cause you used it)
